### PR TITLE
dockerfile enhancements

### PR DESCRIPTION
**Pull request by wjcunningham7**
- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation and CHANGELOG accordingly.
- [ ] I have read the CONTRIBUTING document.

### CODE REVIEW

**Comment by wjcunningham7** [comment]
The purpose of this enhancement is:
- Avoid shipping the server when only the SDK is required (e.g. in executor base images)
- Use a single Dockerfile for SDK and server containers, despite having quite different behaviors
- Allow builds from local source, a Github commit SHA, or a PyPI release

## Sample Build Instructions
Build a SDK-only container from a local source
```
docker build --build-arg BUILD_DATE=`date -u +%Y-%m-%d` --build-arg BUILD_VERSION=0.209.1-rc.0 --tag covalent-sdk:0.209.1-rc.0 .
```

Build a Covalent Server container from a commit SHA checked into this repo:
```
docker build --build-arg COVALENT_SOURCE=sha --build-arg COVALENT_INSTALL_TYPE=server --build-arg COVALENT_COMMIT_SHA=abcdef --build-arg BUILD_DATE=`date -u +%Y-%m-%d` --build-arg BUILD_VERSION=0.209.1-rc.0 --tag covalent-server:abcdef .
```

Build a developer environment (includes build tools but not Covalent itself):
```
docker build --tag covalent-build:latest --target build_server .
```

Build a developer environment, including Covalent:
```
docker build --build-arg COVALENT_INSTALL_TYPE=server --tag covalent-dev:latest --target covalent_install .
```

Finally, also note that the base image can now be customized by providing the additional build argument `--build-arg BASE_IMAGE=<my_base_image>`, with the only requirement that the base image include a compatible version of Python.

Run the server:
```
docker run -p 48008:48008 -d covalent-server:latest
```

File: Dockerfile [hunk 1]
```diff
@@ -18,49 +19,183 @@
...
-COPY --from=build /usr/bin/wget /usr/bin/wget
-COPY --from=build /usr/lib/x86_64-linux-gnu/libpcre2-8.so.0 /usr/lib/x86_64-linux-gnu/libpcre2-8.so.0
-COPY --from=build /usr/lib/x86_64-linux-gnu/libpsl.so.5 /usr/lib/x86_64-linux-gnu/libpsl.so.5
+RUN <<EOL
+  apt-get update
+  apt-get install -y --no-install-recommends \
+    git \
+    rsync \
+    unzip \
+    vim \
```
Comment (venkatBala) [inline Dockerfile:51]: @wjcunningham7 do we need `vim`?

File: Dockerfile [hunk 2]
```diff
@@ -18,49 +19,183 @@
...
+EXPOSE ${COVALENT_SERVER_PORT}
+
+HEALTHCHECK CMD wget --no-verbose --tries=1 --spider http://localhost:${COVALENT_SVC_PORT} || exit 1
+
+RUN sed -i "s#$BUILDROOT#$INSTALLROOT#" $INSTALLROOT/.venv/bin/covalent
+
+#ENTRYPOINT [ "covalent" ]
+ENTRYPOINT [ "/bin/bash" ]
+
+CMD [ "-c", "covalent start --port ${COVALENT_SVC_PORT} && tail -f /dev/null" ]
```
Comment (venkatBala) [inline Dockerfile:199]: we should support the DASK environment variables aka. `COVALENT_NUM_WORKERS`, `COVALENT_MEM_PER_WORKER` & `COVALENT_THREADS_PER_WORKER`, `COVALENT_DEBUG_MODE` ...

`covalent start --workers $COVALENT_NUM_WORKERS --mem-per-worker $COVALENT_MEM_PER_WORKER --threads-per-worker $COVALENT_THREADS_PER_WORKER -d`

**Review by venkatBala — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (wjcunningham7) [reply Dockerfile:51]: For a build/dev container yes

File: Dockerfile (hunk 2 above)
Comment (wjcunningham7) [reply Dockerfile:199]: Agreed, let's expose all the variables here and also find a spot in documentation for them

**Comment by cjao** [comment]
@wjcunningham7 do you think the examples you gave could be added to CONTRIBUTING.md? I for one will likely refer to them when I try the container builds.

**Comment by cjao** [comment]
@wjcunningham7 do you think the examples you gave could be added to CONTRIBUTING.md? I for one will likely refer to them when I try out the containers.

**Comment by wjcunningham7** [comment]
> @wjcunningham7 do you think the examples you gave could be added to CONTRIBUTING.md? I for one will likely refer to them when I try out the containers.

@cjao good suggestion, it has been added

**Comment by wjcunningham7** [comment]
> > @wjcunningham7 do you think the examples you gave could be added to CONTRIBUTING.md? I for one will likely refer to them when I try out the containers.
> 
> @cjao good suggestion, it has been added

https://github.com/AgnostiqHQ/covalent/blob/dockerfile-enhancements/CONTRIBUTING.md#covalent-with-docker

**Comment by wjcunningham7** [comment]
Note: This should not be merged until the CI is tested against it

**Review by venkatBala — DISMISSED** [review]

**Pull request merged**